src/theory/sep/theory_sep.cpp

```diff
@@ -312,219 +278,250 @@ void TheorySep::presolve() {
+      if (slbl != ss)
+      {
+        conc = slbl.eqNode(ss);
+      }
+      // note semantics of sep.nil is enforced globally
+    }
+    else if (satom.getKind() == SEP_EMP)
+    {
+      // conc = slbl.eqNode(
+      // nm->mkConst(EmptySet(slbl.getType())) );
+      Node lem;
+      Node emp_s = nm->mkConst(EmptySet(slbl.getType()));
+      if (polarity)
+      {
+        lem = nm->mkNode(OR, fact.negate(), slbl.eqNode(emp_s));
+      }
+      else
+      {
+        Node kl = nm->mkSkolem("loc", getReferenceType(satom));
+        Node kd = nm->mkSkolem("data", getDataType(satom));
+        Node econc = nm->mkNode(
+            SEP_LABEL,
+            nm->mkNode(SEP_STAR, nm->mkNode(SEP_PTO, kl, kd), d_true),
+            slbl);
+        // Node econc = nm->mkNode( AND, slbl.eqNode( emp_s ).negate(),
+        lem = nm->mkNode(OR, fact.negate(), econc);
+      }
+      Trace("sep-lemma") << "Sep::Lemma : emp : " << lem << std::endl;
+      d_out->lemma(lem);
+    }
+    else
+    {
+      // labeled emp should be rewritten
+      Assert(false);
+    }
+    d_red_conc[slbl][satom] = conc;
+  }
+  if (conc.isNull())
+  {
+    Trace("sep-lemma-debug")
+        << "Trivial conclusion, do not add lemma." << std::endl;
+    return;
   }
+  bool use_polarity = satom.getKind() == SEP_WAND ? !polarity : polarity;
+  if (!use_polarity)
+  {
+    // introduce guard, assert positive version
+    Trace("sep-lemma-debug")
+        << "Negated spatial constraint asserted to sep theory: " << fact
+        << std::endl;
+    Node g = nm->mkSkolem("G", nm->booleanType());
+    d_neg_guard_strategy[g].reset(new DecisionStrategySingleton(
+        "sep_neg_guard", g, getSatContext(), getValuation()));
+    DecisionStrategySingleton* ds = d_neg_guard_strategy[g].get();
+    getDecisionManager()->registerStrategy(DecisionManager::STRAT_SEP_NEG_GUARD,
+                                           ds);
+    Node lit = ds->getLiteral(0);
+    d_neg_guard[slbl][satom] = lit;
+    Trace("sep-lemma-debug")
+        << "Neg guard : " << slbl << " " << satom << " " << lit << std::endl;
+    AlwaysAssert(!lit.isNull());
+    d_neg_guards.push_back(lit);
+    d_guard_to_assertion[lit] = satom;
+    // Node lem = nm->mkNode( EQUAL, lit, conc );
+    Node lem = nm->mkNode(OR, lit.negate(), conc);
+    Trace("sep-lemma") << "Sep::Lemma : (neg) reduction : " << lem << std::endl;
+    d_out->lemma(lem);
+  }
+  else
+  {
+    // reduce based on implication
+    Node lem = nm->mkNode(OR, fact.negate(), conc);
+    Trace("sep-lemma") << "Sep::Lemma : reduction : " << lem << std::endl;
+    d_out->lemma(lem);
+  }
+}
 
-  if( e == EFFORT_LAST_CALL && !d_conflict && !d_valuation.needCheck() ){
+bool TheorySep::isSpatialKind(Kind k) const
+{
+  return k == SEP_STAR || k == SEP_WAND || k == SEP_PTO || k == SEP_EMP;
+}
+
+void TheorySep::postCheck(Effort level)
+{
+  if (level == EFFORT_LAST_CALL && !d_state.isInConflict()
+      && !d_valuation.needCheck())
+  {
```

Comment:
Can we refactor the function to use less indentation by doing something like:

```
if (level != EFFORT_LAST_CALL || d_state.isInConflict() || d_valuation.needCheck())
{
  return;
}
```

Reply:
Sure, was trying to avoid some indentation changes to keep the LOC changes small but I can change this.